src/Adapter/Image/ImageGenerator.php

```diff
@@ -0,0 +1,105 @@
+<?php
+/**
+ * Copyright since 2007 PrestaShop SA and Contributors
+ * PrestaShop is an International Registered Trademark & Property of PrestaShop SA
+ *
+ * NOTICE OF LICENSE
+ *
+ * This source file is subject to the Open Software License (OSL 3.0)
+ * that is bundled with this package in the file LICENSE.md.
+ * It is also available through the world-wide-web at this URL:
+ * https://opensource.org/licenses/OSL-3.0
+ * If you did not receive a copy of the license and are unable to
+ * obtain it through the world-wide-web, please send an email
+ * to license@prestashop.com so we can send you a copy immediately.
+ *
+ * DISCLAIMER
+ *
+ * Do not edit or add to this file if you wish to upgrade PrestaShop to newer
+ * versions in the future. If you wish to customize PrestaShop for your
+ * needs please refer to https://devdocs.prestashop.com/ for more information.
+ *
+ * @author    PrestaShop SA and Contributors <contact@prestashop.com>
+ * @copyright Since 2007 PrestaShop SA and Contributors
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ */
+
+declare(strict_types=1);
+
+namespace PrestaShop\PrestaShop\Adapter\Image;
+
+use Configuration;
+use ImageManager;
+use PrestaShop\PrestaShop\Core\Image\Uploader\Exception\ImageOptimizationException;
+use PrestaShop\PrestaShop\Core\Image\Uploader\Exception\ImageUploadException;
+use PrestaShopException;
+
+class ImageGenerator
+{
+    /**
+     * @param string $imagePath
+     * @param array $imageTypes
+     *
+     * @return bool
+     *
+     * @throws ImageOptimizationException
+     * @throws ImageUploadException
+     */
+    public function generateImagesByTypes(string $imagePath, array $imageTypes): bool
+    {
+        $resized = true;
+
+        try {
+            foreach ($imageTypes as $imageType) {
+                $resized &= $this->resize($imagePath, $imageType);
+            }
+        } catch (PrestaShopException $e) {
+            //@todo: this exception extends generic Exception and is in Uploader ns.
+            //@todo: could it extend new Core/Image/ImageException? this would introduce BC break. (make the same with other Uploader/Exception's ?)
+            throw new ImageOptimizationException('Unable to resize one or more of your pictures.');
```

Comment:
@jolelievre @matks 
Its a bit of a mess in https://github.com/PrestaShop/PrestaShop/tree/develop/src/Core/Image/Uploader/Exception
do you think we could clean it somehow with as less BC breaks as possible?
I think all those exceptions should extend ImageException (which extends core) & ImageOptimizationException could at least be moved one dir out to Core/Image/Exception. 
Or maybe all of those exceptions could simply stay in Core/Image/Exception ? (im worried, because it is messy & not following any convention, so I don't want to use those anywhere)

Reply:
I agree that they should extend `ImageException` at least, and this wouldn't break anything since it respects their initial ancestor and only adds a "layer" of exception

However moving them might be a little trickier as it would change their namespace and create a BC break, besides having sub exceptions in sub namespaces is not that messy We did that a lot for product exceptions in `namespace PrestaShop\PrestaShop\Core\Domain\Product\AttributeGroup\Exception;` `namespace PrestaShop\PrestaShop\Core\Domain\Product\Combination\Exception;`

For domains that can potentially have many types of exceptions it makes sense, and I'm sure Image will have other potential sub domains, including generation or manipulation So maybe you can create a new one however dedicated to this sub domain

Reply:
Actually I realize one of those exceptions should indeed be moved `ImageOptimizationException` would be better placed in a Generator or Manipulator image sub domain

Reply:
done. Described BC breaks in PR description.